### ISSUE TITLE
Watch mode not working for many files in a large project

### ISSUE THREAD

**Issue by szucs2020** [webpack/webpack#10339]
# Bug report

**What is the current behavior?**
Webpack recompiles when saving some files in my project but not others when running in watch mode. Other than the fact that they are all a mix of .ts .tsx and .less files, there does not seem to be anything relating the files affected. They are in separate folders, and in some cases files next to each other within folders are affected differently. The issue is random but consistent as far as which files are affected. 

To eliminate the possibility that there was some file corruption, I completely wiped and cloned our repo. This did not fix the issue, and the same files that were affected before continued to be affected.

To debug the issue, I was putting console.log statements in the `watchpack` source. I found that the on change callbacks from chokidar were not being called, and when I added a callback for the "all" / "raw" events, there were only messages for the files that caused a recompile.

It's possible this is a `chokidar` issue, but I don't know enough about how it works to say for sure. I downloaded `chokidar-cli` and `filemon` to test the `chokidar` file watching and check to make sure the issue wasn't with `FSEvents`. While watching my project, the events all appeared correctly for the files that were not causing recompiles with webpack.

**If the current behavior is a bug, please provide the steps to reproduce.**
Unfortunately I can't find anyone else at my company who has this issue with webpack on our project, so I can't give you much else other than that I save a file and webpack recompiles, then when I save on an affected file it does not.

I'm sorry but at this time I can't give you the configuration or any code. If it makes this ticket useless I apologize but I don't believe I'm allowed to give out any of our code. If you have any questions about the configuration I can answer them though. I've tried various watch mode settings like polling to no avail.

**What is the expected behavior?**
When webpack is run in watch mode, saving a file thats part of the project should trigger a recompile. The files I'm working with are react Typescript .tsx files and regular Typescript .ts files.

**Other relevant information:**
webpack version:  4.29.6
Node.js version: 12.14.1
Operating System: macOS Mojave 10.14.6 
Additional tools:

**Comment by evilebottnawi** [comment]
Please create reproducible test repo

**Comment by szucs2020** [comment]
okay, I'll try and get something together as small as I can. Its proving to be pretty difficult because it seems like the project size has something to do with it, but I'll get some more info either way.

**Comment by sokra** [comment]
webpack always watches the folders containing the files. So if you notice that there are two files in one folder, where one is affected and the other is not, it's probably not a watching issue, but a issue of undeclared dependencies on these files.

Does this happen only on certain file types or is that unrelated? Always the same files or random? Any special configuration like `resolve.symlinks: false`?

**Comment by szucs2020** [comment]
So finding a minimum project example has been really hard, and I figured out the reason why. The dependencies are complicated, so to reduce to a minimum I would only import one type from the broken files to my bootstrap file to try and delete all the others. Basically what I discovered was that by directly importing a type from these broken files, it actually fixed the issue. The files are definitely used in the project, but more indirectly. 

`resolve.symlinks: false` is not in my config file.

**Comment by sokra** [comment]
> Basically what I discovered was that by directly importing a type from these broken files, it actually fixed the issue.

Is the issue related to typescript and type-only imports?

Type-only imports do not create a dependency visible to webpack, so the type-exporting module is not visible to webpack and it's the responsiblity of the typescript loader to declare this dependency.

**Comment by szucs2020** [comment]
Sorry to take so long to close this, I sorted it out a couple weeks ago and forgot to update you. Basically there were two issues, and neither were related to webpack. The first, was that my repository folder was corrupted, which I learned sometimes happens on OSX. I fixed it by recloning our repo. After fixing the corrupted folder, I still had a few files left that didn't work, but that was because there were entire folders worth of code in my company's repo that were unused. They have now all been deleted. Woops...
